eZ/Publish/API/Repository/Iterator/BatchIterator.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Iterator;
+
+use Iterator;
+
+final class BatchIterator implements Iterator
+{
+    public const DEFAULT_BATCH_SIZE = 25;
+
+    /** @var \eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter */
+    private $adapter;
+
+    /** @var Iterator|null */
+    private $innerIterator;
+
+    /** @var int */
+    private $batchSize;
+
+    /** @var int */
+    private $position;
+
+    public function __construct(
+        BatchIteratorAdapter $adapter,
+        int $batchSize = self::DEFAULT_BATCH_SIZE
+    ) {
+        $this->adapter = $adapter;
+        $this->batchSize = $batchSize;
+        $this->position = 0;
+    }
+
+    public function current()
+    {
+        if (!$this->isInitialized()) {
+            $this->initialize();
+        }
+
+        return $this->innerIterator->current();
+    }
+
+    public function next(): void
+    {
+        if (!$this->isInitialized()) {
+            $this->initialize();
+        }
+
+        ++$this->position;
+        $this->innerIterator->next();
+        if (!$this->innerIterator->valid() && ($this->position % $this->batchSize) === 0) {
```

Comment:
This seems a little odd. The second check assumes that in some cases the `Iterator` supplied by the `adapter::fetch` method may contain more items than batch size?

Why do we even care about it? If we somehow ended up with Iterator that contains more items than it's supposed to, we would continue iterating until it's exhausted - and then we would simple fetch using the `position` property, just like it's already done.

Reply:
> The second check assumes that in some cases the Iterator supplied by the adapter::fetch method may contain more items than batch size?

No, `($this->position % $this->batchSize) === 0` allows to avoid additional `\eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter::fetch` call (in average case) when we reach last item (number of items on last page < batch size).

Reply:
Makes sense. Now I get it. :+1: